src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -657,6 +697,72 @@ int getBufferedLength() {
             return (ibuffer == null ? 0 : ibuffer.size());
         }
 
+        /**
+         *  ByteBuffer wrapper for intrinsic implGCMCrypt
+         */
+        int implGCMCrypt(GCMOperation op, ByteBuffer src, ByteBuffer dst) {
+            int srcLen = src.remaining() - (src.remaining() % blockSize);
+            int rlen = srcLen;
+
+
+            if (srcLen < blockSize) {
+                return 0;
+            }
+
+            // 'in' and 'out' are always set together, just need to check 'in'
+            if (in == null || (in.length != PARALLEL_LEN
+                && in.length < srcLen)) {
+                in = new byte[Math.min(PARALLEL_LEN, srcLen)];
+                out = new byte[Math.min(PARALLEL_LEN, srcLen)];
+            }
+
+            byte[] ct;
+            if (encryption) {
+                ct = out;
+            } else {
+                ct = in;
+            }
+
+            if (srcLen > PARALLEL_LEN) {
+
+                if (src.hasArray() && dst.hasArray()) {
+                    int l = rlen;
+
+                    if (encryption) {
+                        rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
+                            src.arrayOffset() + src.position(), src.remaining(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    } else {
+                        rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
+                            src.arrayOffset() + src.position(), src.remaining(),
+                            src.array(), src.arrayOffset() + src.position(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    }
+                    src.position(src.position() + srcLen);
+                    dst.position(dst.position() + srcLen);
+                } else {
+
+                    do {
+                        src.get(in, 0, PARALLEL_LEN);
+                        rlen -= GaloisCounterMode.implGCMCrypt(in, 0,
+                            PARALLEL_LEN, ct, 0, out, 0, gctr, ghash);
+                        dst.put(out, 0, PARALLEL_LEN);
+                    } while (rlen > PARALLEL_LEN);
+                }
+            }
+            if (rlen >= blockSize) {
+                src.get(in, 0, rlen);
+                rlen = op.update(in, 0, rlen, out, 0);
+            }
+
+            dst.put(out, 0, rlen);
+            processed += srcLen;
```

Comment:
It seems that callers of this implGCMCrypt() method such as GCMEngine.doLastBlock() adds the returned value to the "processed" field which looks like double counting? However, some caller such as GCMEncrypt.doUpdate() does not. Seems inconsistent and may lead to wrong value for the "processed" field?

Reply:
All the callers that use GCMOperations, ie op.update(...), have the processed value updated.  implGCMCrypt() calls op.update() and updates the value.  It cannot double count 'processed' is not updated after implGCMCrypt().  I can see your point, but the other methods do not have access to 'processed' and would mean I copy that line 3 times elsewhere.  I'd rather keep it as is

Reply:
As long the "processed" value is correct, it's fine.
Not sure if I may be missing some subtle things, GCMEngine.implGCMCrypt(GCMOperation op, ByteBuffer src, ByteBuffer dst) impl would increment "processed" with "srcLen". But then in GCMEngine.doLastBlock(GCMOperation op, ByteBuffer buffer, ByteBuffer src, ByteBuffer dst) impl, it calls the GCMEngine.implGCMCrypt(GCMOperation op, ByteBuffer src, ByteBuffer dst) method and stores the return value into "resultLen" and then again increment "processed" with "resultLen" after op.doFinal(...) call. Since "resultLen" contains the number of bytes processed by GCMEngine.implGCMCrypt(GCMOperation op, ByteBuffer src, ByteBuffer dst) method already, adding it to "prcessed" looks like double counting. Not sure what did I miss.